contracts/utils/cryptography/draft-ERC7739Utils.sol

```diff
@@ -0,0 +1,282 @@
+     *      keccak256(
+     *          abi.encode(
+     *              NESTED_TYPED_DATA_TYPEHASH(contentsType), // See {NESTED_TYPED_DATA_TYPEHASH}
+     *              contents,
+     *              // See {IERC5267-eip712Domain} for the following arguments from the verifying contract's domain
+     *              keccak256(bytes(name)),
+     *              keccak256(bytes(version)),
+     *              chainId,
+     *              verifyingContract,
+     *              salt,
+     *              keccak256(abi.encodePacked(extensions))
+     *          )
+     *      )
+     *  )
+     *);
+     *
+     * NOTE: The arguments should be those of the verifying application. See {EIP712-_domainSeparatorV4}
+     * and {IERC5267-eip712Domain} for more details of how to obtain these values. Respectively, they
+     * must be obtained from the verifying contract (e.g. an Account) and the application domain.
+     */
+    function toNestedTypedDataHash(
+        bytes32 separator,
+        bytes32 hashedTypedDataNestedStruct
+    ) internal pure returns (bytes32) {
+        return MessageHashUtils.toTypedDataHash(separator, hashedTypedDataNestedStruct);
+    }
+
+    /**
+     * @dev Computes the nested EIP-712 type hash for the given contents type.
+     *
+     * The `contentsTypeName` is the string name in the app's domain before the parentheses
+     * (e.g. Transfer in `Transfer(address to,uint256 amount)`).
+     *
+     * ```solidity
+     * TypedDataSign({contentsTypeName},bytes1 fields,string name,string version,uint256 chainId,address verifyingContract,bytes32 salt,uint256[] extensions){contentsType}
+     * ```
+     *
+     * Requirements:
+     *  - `contentsType` must be a valid EIP-712 type (see {tryValidateContentsType})
+     */
+    // solhint-disable-next-line func-name-mixedcase
+    function NESTED_TYPED_DATA_TYPEHASH(bytes calldata contentsType) internal pure returns (bytes32) {
+        (bool valid, bytes calldata contentsTypeName) = tryValidateContentsType(contentsType);
+        if (!valid) revert InvalidContentsType();
+        return NESTED_TYPED_DATA_TYPEHASH(contentsType, contentsTypeName);
+    }
+
+    /**
+     * @dev Same as {NESTED_TYPED_DATA_TYPEHASH} but with the `contentsTypeName` already validated.
+     *
+     * See {tryValidateContentsType}.
+     */
+    // solhint-disable-next-line func-name-mixedcase
+    function NESTED_TYPED_DATA_TYPEHASH(
+        bytes calldata contentsType,
+        bytes calldata contentsTypeName
+    ) internal pure returns (bytes32) {
+        return
+            keccak256(
+                abi.encodePacked(
+                    "TypedDataSign(",
+                    contentsTypeName,
+                    "bytes1 fields,",
+                    "string name,",
+                    "string version,",
+                    "uint256 chainId,",
+                    "address verifyingContract,",
+                    "bytes32 salt,",
+                    "uint256[] extensions",
```

Comment:
Currently this seems like it's doing a lot of concatenation at runtime that can be done statically.

```suggestion
                    "bytes1 fields,string name,string version,uint256 chainId,address verifyingContract,bytes32 salt,uint256[] extensions",
```